Add a error check in postHijacked to avoid docker exec command blocking.

keloyang opened the pull request:
When user execute docker exec command(client.ContainerExecAttach), docker daemon maybe have err return because
of ExecExists check, and then the hijack stream will not be close, it can lead to docker exec command block.

Signed-off-by: yangshukui <yangshukui@huawei.com>

**- What I did**
Add a error check in postHijacked to avoid docker exec command blocking.

**- How to verify it**

with the follown steps， the problem can be reproduced easyly.
1. start a testone container
```
docker run -d --name testone ubuntu bash -c 'for (( i = 1; i <= 200000; i++ )) do echo -n "$i";sleep 1;done'
```
2. run pause.sh
```
#!/bin/bash
while true; do
        docker pause testone
        docker unpause testone
done
```
3. run exec.sh
```
#!/bin/bash
while true; do
        echo "exec ${i}th times"
        docker exec -ti testone bash  -c "sleep 0.5 ;exit"
        ((i++))
done
```
just wait a minute, you will see the blocking.

you can also modify daemon'code to help reproduce like this:
```
--- a/api/server/router/container/exec.go
+++ b/api/server/router/container/exec.go
@@ -77,7 +77,7 @@ func (s *containerRouter) postContainerExecStart(ctx context.Context, w http.Res
        if err := json.NewDecoder(r.Body).Decode(execStartCheck); err != nil {
                return err
        }
-
+       time.Sleep(2 * time.Second)
        if exists, err := s.backend.ExecExists(execName); !exists {
                return err
        }
```
then run " docker unpause testone; sleep 2;docker pause testone" in a shell, and run " docker exec -ti testone bash  -c "sleep 0.5 ;exit"" in another shell immediately.

keloyang commented:
ping @dnephin

coolljt0725 commented:
ping @vdemeester  @thaJeztah

cpuguy83 commented on client/hijack.go:
```diff
@@ -72,7 +74,21 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
+	if err != nil {
+		return types.HijackedResponse{}, err
+	}
+	if resp != nil {
```
This nil check looks like it's not needed.

thaJeztah replied:
Yes, I think it's cleaner without @keloyang can you remove it?

keloyang replied:
It's true that I need remove it , thank you.

cpuguy83 reviewed (approved):
LGTM

dnephin commented on client/hijack.go:
```diff
@@ -72,7 +74,21 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
+			}
+			return types.HijackedResponse{}, fmt.Errorf("Error response from daemon: %s", bytes.TrimSpace(body))
+		}
+	}
```
What do you think about something like this:

```go

defer ensureReaderClosed(resp)

switch resp.StatusCode {
case http.StatusOK, http.StatusSwitchingProtocols:
    rwc, br := clientconn.Hijack()
    return types.HijackedResponse{Conn: rwc, Reader: br}, nil
}

body, err := ioutil.ReadAll(resp.Body)
...
return types.HijackedResponse{}, fmt.Errorf("...")
```

I think you want the response closed called in both cases?

keloyang replied:
@dnephin ，thanks for your review and advice. 
I just want the response close called when resp.StatusCode != http.StatusOK(or http.StatusSwitchingProtocols)， If  resp.StatusCode is http.StatusSwitchingProtocols, do we must close the resp ? I'm not sure.

keloyang commented:
ping @thaJeztah

keloyang commented:
ping @thaJeztah @dnephin @aaronlehmann

thaJeztah commented on client/hijack.go:
```diff
@@ -72,11 +74,23 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
-
-	rwc, br := clientconn.Hijack()
+	resp, err := clientconn.Do(req)
+	if resp != nil {
```
I think the previous comment https://github.com/moby/moby/pull/33262#discussion_r117484500 meant to say that we don't have to check `resp` if `err != nil`; we can still return early here, to reduce complexity, e.g.;

```go
resp, err := clientconn.Do(req)
if err != nil {
	return types.HijackedResponse{}, err
}

defer resp.Body.Close()

switch resp.StatusCode {
case http.StatusOK, http.StatusSwitchingProtocols:
	rwc, br := clientconn.Hijack()
	return types.HijackedResponse{Conn: rwc, Reader: br}, err
}
body, err := ioutil.ReadAll(resp.Body)
if err != nil {
	return types.HijackedResponse{}, err
}
return types.HijackedResponse{}, fmt.Errorf("Error response from daemon: %s", bytes.TrimSpace(body))
```

thaJeztah reviewed (changes requested):
left one suggestion, but LGTM otherwise

ping @dnephin PTAL as well

coolljt0725 commented:
@keloyang  squash your commits

keloyang commented:
@coolljt0725 , I apologize and update the pr. Thank you.

Commit:
Add a error check in postHijacked to avoid docker exec command blocking.

When user execute docker exec command, docker daemon maybe have err return because
of ExecExists check, and then the hijack stream will not be close, it can lead to
docker exec command block.

Signed-off-by: yangshukui <yangshukui@huawei.com>

dnephin reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM, thanks!

tonistiigi commented on client/hijack.go:
```diff
@@ -72,11 +74,23 @@ func (cli *Client) postHijacked(ctx context.Context, path string, query url.Valu
-	rwc, br := clientconn.Hijack()
+	defer resp.Body.Close()
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusSwitchingProtocols:
```
@keloyang Why is `StatusOK` allowed here?

keloyang commented:
```
func (s *containerRouter) postContainersAttach(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
        err := httputils.ParseForm(r)
        if err != nil {
                return err
        }
        containerName := vars["name"]

        _, upgrade := r.Header["Upgrade"]
        detachKeys := r.FormValue("detachKeys")

        hijacker, ok := w.(http.Hijacker)
        if !ok {
                return fmt.Errorf("error attaching to container %s, hijack connection missing", containerName)
        }

        setupStreams := func() (io.ReadCloser, io.Writer, io.Writer, error) {
                conn, _, err := hijacker.Hijack()
                if err != nil {
                        return nil, nil, nil, err
                }

                // set raw mode
                conn.Write([]byte{})

                if upgrade {
                        fmt.Fprintf(conn, "HTTP/1.1 101 UPGRADED\r\nContent-Type: application/vnd.docker.raw-stream\r\nConnection: Upgrade\r\nUpgrade: tcp\r\n\r\n")
                } else {
                        fmt.Fprintf(conn, "HTTP/1.1 200 OK\r\nContent-Type: application/vnd.docker.raw-stream\r\n\r\n")
                }
...
```

tonistiigi commented:
@keloyang But this function always sends the `upgrade` header so it should not reach the second condition.

keloyang commented:
@tonistiigi thanks. I have found that out, if we don't  allowe StatusOK  in postHijack，we should remove 'fmt.Fprintf(conn, "HTTP/1.1 200 OK\r\nContent-Type: application/vnd.docker.raw-stream\r\n\r\n")' from docker daemon, WDYT ?
and ping @dnephin @thaJeztah

tonistiigi commented:
It may be backward compatibility with older clients(didn't confirm), but shouldn't be a problem for the client pkg as we didn't support new clients with old daemon until very recently.

keloyang commented:
Do we have any problems to allow  StatusOK in postHijack ?
